llvm/include/llvm/IR/InlineAsm.h

```diff
@@ -292,101 +274,152 @@ class InlineAsm final : public Value {
+    /// that this input operand is tied to a previous output operand.
+    void setMatchingOp(unsigned MatchedOperandNo) {
+      assert(MatchedOperandNo <= 0x7fff && "Too big matched operand");
+      assert(GrabBag == 0 && "Matching operand already set");
+      GrabBag = MatchedOperandNo;
+      IsMatched = true;
+    }
 
-  /// getNumOperandRegisters - Extract the number of registers field from the
-  /// inline asm operand flag.
-  static unsigned getNumOperandRegisters(unsigned Flag) {
-    return (Flag & 0xffff) >> 3;
-  }
+    /// setRegClass - Augment an existing flag with the required register class
+    /// for the following register operands. A tied use operand cannot have a
+    /// register class, use the register class from the def operand instead.
+    void setRegClass(unsigned RC) {
+      // Store RC + 1, reserve the value 0 to mean 'no register class'.
+      ++RC;
+      assert(!isImmKind() && "Immediates cannot have a register class");
+      assert(!isMemKind() && "Memory operand cannot have a register class");
+      assert(RC <= 0x7fff && "Too large register class ID");
+      assert(GrabBag == 0 && "Register class already set");
+      GrabBag = RC;
+    }
 
-  /// isUseOperandTiedToDef - Return true if the flag of the inline asm
-  /// operand indicates it is an use operand that's matched to a def operand.
-  static bool isUseOperandTiedToDef(unsigned Flag, unsigned &Idx) {
-    if ((Flag & Flag_MatchingOperand) == 0)
-      return false;
-    Idx = (Flag & ~Flag_MatchingOperand) >> 16;
-    return true;
-  }
+    /// Augment an existing flag with the constraint code for a memory
+    /// constraint.
+    void setMemConstraint(unsigned Constraint) {
+      assert((isMemKind() || isFuncKind()) &&
+             "InputFlag is not a memory or function constraint!");
+      assert(Constraint <= 0x7fff && "Too large a memory constraint ID");
+      assert(Constraint <= Constraints_Max && "Unknown constraint ID");
+      assert(GrabBag == 0 && "Mem constraint already set");
+      GrabBag = Constraint;
+    }
 
-  /// hasRegClassConstraint - Returns true if the flag contains a register
-  /// class constraint.  Sets RC to the register class ID.
-  static bool hasRegClassConstraint(unsigned Flag, unsigned &RC) {
-    if (Flag & Flag_MatchingOperand)
-      return false;
-    unsigned High = Flag >> 16;
-    // getFlagWordForRegClass() uses 0 to mean no register class, and otherwise
-    // stores RC + 1.
-    if (!High)
-      return false;
-    RC = High - 1;
-    return true;
-  }
+    // TODO: do we even need this? WTF is it doing?
+    // IIUC, It was masking off the bottom 22b.
+    // ~(0x7fff << 16) == 0x3FFFFF
+    void convertMemFlagWordToMatchingFlagWord() {
+      assert(isMemKind());
+      GrabBag = 0;
+      // GrabBag &= 0x3F;
+      // GrabBag = GrabBag << Constraints_ShiftAmount; // TODO: is this even correct?
+      // return InputFlag & ~(0x7fff << Constraints_ShiftAmount);
+    }
```

Comment:
This should get cleaned up before landing; I'm not quite sure what it was even doing in the past.

Reply:
It was added in 60f1db0525003. I'm not sure what it does either...I guess the masked stuff converts from mem to matching?

Reply:
ah, I did my math wrong (different results for 64b vs 32b). `~(0x7fff << Constraints_ShiftAmount)` == 2147549183 == 0x8000FFFF.

So it's using masking to:
1. retain isMatched
2. retain num operands
3. retain kind
4. reset/clear data.

Let me fix this+retest.

Reply:
done in https://github.com/llvm/llvm-project/pull/65649/commits/c9b90690d8d95d562be3b74568c44fdf351ca5b9